packages/xstate-store/src/atom.ts

```diff
@@ -1,85 +1,221 @@
-import { toObserver } from './toObserver';
 import {
-  AnyAtom,
-  Atom,
-  Observer,
-  Readable,
-  ReadonlyAtom,
-  Subscription
-} from './types';
+  createReactiveSystem,
+  Link,
+  Subscriber,
+  SubscriberFlags
+} from './alien';
+import { toObserver } from './toObserver';
+import { Atom, AtomOptions, Observer, Readable, ReadonlyAtom } from './types';
+
+const {
+  link,
+  propagate,
+  endTracking,
+  startTracking,
+  updateDirtyFlag,
+  processComputedUpdate,
+  processEffectNotifications
+} = createReactiveSystem({
+  updateComputed(computed: ReadonlyAtom<any>) {
+    return computed.update();
+  },
+  notifyEffect(effect: Effect) {
+    effect.notify();
+    return true;
+  }
+});
+
+let activeSub: Subscriber | undefined = undefined;
 
 export function createAtom<T>(
-  getValue: (read: <U>(atom: Readable<U>) => U) => T
+  getValue: (read: <U>(atom: Readable<U>) => U) => T,
+  options?: AtomOptions<T>
 ): ReadonlyAtom<T>;
-export function createAtom<T>(initialValue: T): Atom<T>;
 export function createAtom<T>(
-  valueOrFn: T | ((read: <U>(atom: Readable<U>) => U) => T)
+  initialValue: T,
+  options?: AtomOptions<T>
+): Atom<T>;
+export function createAtom<T>(
+  valueOrFn: T | ((read: <U>(atom: Readable<U>) => U) => T),
+  options?: AtomOptions<T>
 ): Atom<T> | ReadonlyAtom<T> {
-  const current = { value: undefined as T };
-  let observers: Set<Observer<T>> | undefined;
-
-  // Handle computed case
   if (typeof valueOrFn === 'function') {
-    const subs = new Map<AnyAtom, Subscription>();
-    let observedAtoms = new Set<AnyAtom>();
-
-    const getValue = valueOrFn as (read: <U>(atom: Atom<U>) => U) => T;
-    const read = (atom: AnyAtom) => {
-      observedAtoms.add(atom);
-      const val = atom.get();
-      if (subs.has(atom)) {
-        return val;
+    return computed(valueOrFn as () => T, options);
+  }
+
+  // Create plain object atom
+  const atom = {
+    currentValue: valueOrFn,
+    options,
+    // Dependency fields
+    subs: undefined as Link | undefined,
```

Comment:
I feel like this abuses casts and makes it harder to determine how those concepts compose here. The `atom`'s declaration should get a type annotation here (and the same thing applies to other objects that use this pattern in this file

Reply:
Cleaned this up; thanks for pointing it out